cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +555,308 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+      auto& schema_elem = per_file_metadata[0].schema[schema_idx];
+
+      // ensure equal number of children first to avoid any segfaults in children
+      if (schema_elem.num_children == static_cast<int32_t>(arrow_schema.children.size())) {
+        // true if and only if true for all children as well
+        return std::all_of(thrust::make_counting_iterator(0),
+                           thrust::make_counting_iterator(schema_elem.num_children),
+                           [&](auto const& idx) {
+                             return validate_schemas(arrow_schema.children[idx],
+                                                     schema_elem.children_idx[idx]);
+                           });
+      } else {
+        return false;
+      }
+    };
+
+  // Function to co-walk arrow and parquet schemas
+  std::function<void(arrow_schema_data_types&, int)> co_walk_schemas =
+    [&](arrow_schema_data_types& arrow_schema, int schema_idx) {
+      auto& schema_elem = per_file_metadata[0].schema[schema_idx];
+      std::for_each(thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(schema_elem.num_children),
+                    [&](auto const& idx) {
+                      co_walk_schemas(arrow_schema.children[idx], schema_elem.children_idx[idx]);
+                    });
+
+      // true for DurationType columns only for now.
+      if (arrow_schema.type.id() != type_id::EMPTY) {
+        schema_elem.arrow_type = arrow_schema.type.id();
+      }
+    };
+
+  auto schema_root       = get_schema(0);
+  auto arrow_schema_root = arrow_schema.value();
+
+  // verify equal number of children at root level
+  if (schema_root.num_children != static_cast<int32_t>(arrow_schema_root.children.size())) {
+    CUDF_LOG_DEBUG("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // Verify equal number of children at all sub-levels
+  if (not std::all_of(thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(schema_root.num_children),
+                      [&](auto const& idx) {
+                        return validate_schemas(arrow_schema_root.children[idx],
+                                                schema_root.children_idx[idx]);
+                      })) {
+    CUDF_LOG_DEBUG("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // All good, now co-walk schemas
+  std::for_each(thrust::make_counting_iterator(0),
+                thrust::make_counting_iterator(schema_root.num_children),
+                [&](auto const& idx) {
+                  co_walk_schemas(arrow_schema_root.children[idx], schema_root.children_idx[idx]);
+                });
+
+  return;
+}
+
+const void* aggregate_reader_metadata::decode_ipc_message(std::string& serialized_message) const
```

Comment:
```suggestion
void const* aggregate_reader_metadata::decode_ipc_message(std::string& serialized_message) const
```
east const. the same for the rest of the function